src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs

```diff
@@ -128,90 +142,104 @@ public void RelayWorker()
         {
             this.logger.LogTrace("()");
 
-            this.asyncLoop = this.asyncLoopFactory.Run($"{this.name}.RelayWorker", async token =>
+            this.relayWorkerTask = Task.Factory.StartNew(async delegate
```

Comment:
create InitializeAsync which will set rWT, which will be private async Task RelayWorkerAsync
don't use Task.Factory.StartNew, you don't need this at all, it will work:

```
public void Initialize()
{
inlog
    this.relayWorkerTask = this.RelayWorkerAsync();
outlog
}
```

just like that

and also we could migrate `            this.signals.SubscribeForBlocks(this.blockStoreSignaled);` to that initialization method from BSF

Reply:
>this.relayWorkerTask = this.RelayWorkerAsync();

that locks current context because the task does Wait(). So we still need to start a separate task like: 
` this.relayWorkerTask = Task.Run(async delegate { await this.RelayWorkerAsync(); });`

Reply:
But I like the idea of Initialize method, will do.

Reply:
Fixed.

Reply:
no :D tha does not lock current context because no blocking waits are allowed, you only wait with await and that does not block

Reply:
ManualResetEventSlim don't have an async wait method so in order to do it without starting a task we need to add a manual async version of ManualResetEventSlim or use smth else.

Reply:
that is true, but it is possible to construct it
alternatively we can use non-async relay worker which will be based on handle waits

Reply:
we can. But does it worth it?

Reply:
i think it is worth it, but anyway, if you ever want to run task on background, use Task.Run, not factory start new